### PR TITLE
(EAI-252, EAI-253) Ingest & Chunk Code Files

### PR DESCRIPTION
**Pull request by nlarew**
Jira:
- (EAI-252) [Ingest] Code as text files
- (EAI-253) [Ingest] Language-aware code chunking

## Changes

- Adds `CodeOnGithubTextDataSource` to ingest code from a repo in `txt` format
- Adds a partial implementation of MongoDB Developer repos in `mongodbDeveloperCodeRepos`
- Adds language-aware chunking for everything supported by langchain

## Notes

- The repos in `mongodbDeveloperCodeRepos` are mainly just an example for now. We don't actually include them in our list of sources to ingest.
- There is a lot of config if we want to include all these repos. We should instead consider a pull-based approach e.g. https://jira.mongodb.org/browse/EAI-270

### CODE REVIEW

File: packages/ingest-mongodb-public/src/sources/mongodbDeveloperCodeRepos.test.ts
Comment (mongodben) [inline packages/ingest-mongodb-public/src/sources/mongodbDeveloperCodeRepos.test.ts:1]: can you do similar smoke tests for each repo to make sure that we don't loose them?

File: packages/mongodb-rag-ingest/src/sources/CodeOnGithubTextDataSource.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        /\.gcloudignore/,
+        /\.editorconfig/,
+        /\.vscode/,
+        ...(repoLoaderOptions?.ignoreFiles ?? []),
+      ],
+    },
+    async handleDocumentInRepo(document) {
+      const page: Page = {
+        body: document.pageContent,
+        format: "txt",
```
Comment (mongodben) [inline packages/mongodb-rag-ingest/src/sources/CodeOnGithubTextDataSource.ts:47]: why not let this be passable to account for different formats of different languages? 

`java`, `js`, etc

Comment (mongodben) [reply packages/mongodb-rag-ingest/src/sources/CodeOnGithubTextDataSource.ts:47]: supporting "intelligent" chunking on a variety of different languages shouldn't be too hard using the langchain `RecursiveCharacterTextSplitter.fromLanguage()` method.

**Comment by mongodben** [comment]
> The repos in mongodbDeveloperCodeRepos are mainly just an example for now. We don't actually include them in our list of sources to ingest.

what are you thinking for next steps here @nlarew?

**Review by mongodben — CHANGES_REQUESTED** [review]
nice work. see comments. most important thing in my mind is why use the `txt` format instead of for each page, use whatever the file extension is of that page?

File: packages/ingest-mongodb-public/src/sources/mongodbDeveloperCodeRepos.test.ts
Comment (nlarew) [reply packages/ingest-mongodb-public/src/sources/mongodbDeveloperCodeRepos.test.ts:1]: Yup added one for all 6 configured repos

File: packages/mongodb-rag-ingest/src/sources/CodeOnGithubTextDataSource.ts (hunk 1 above)
Comment (nlarew) [reply packages/mongodb-rag-ingest/src/sources/CodeOnGithubTextDataSource.ts:47]: I went a little crazy on this one and added some machinery around a list of "canonical" page formats (and their synonyms). It's a superset of the existing type and should be relatively flexible, though maybe I should add a way to append to it from the ingest config 🤔

**Comment by nlarew** [comment]
> > The repos in mongodbDeveloperCodeRepos are mainly just an example for now. We don't actually include them in our list of sources to ingest.
> 
> what are you thinking for next steps here @nlarew?

@mongodben Once the broader implementation here is LGTM'd I'd like to run an A/B test with our quality evaluation to see if/how the raw code chunks help

File: packages/mongodb-rag-core/src/PageFormat.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  ["kotlin", "kt"],
+  ["latex", "tex"],
+  ["objective-c", "m"],
+  ["php", "php"],
+  ["python", "py"],
+  ["ruby", "rb"],
+  ["rust", "rs"],
+  ["scala", "sc"],
+  ["shell", "sh"],
+  ["swift", "swift"],
```
Comment (mongodben) [inline packages/mongodb-rag-core/src/PageFormat.ts:44]: also can be just 1 elem, no?

File: packages/mongodb-rag-ingest/src/embed/chunkCode.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        : preTransformChunk;
+
+      const chunk = {
+        ...transformedChunk,
+        tokenCount: tokenizer.encode(transformedChunk.text).bpe.length,
+      };
+      // const { metadata } = extractFrontMatter(transformedChunk.text);
+      // if (metadata) {
+      //   chunk["metadata"] = metadata;
+      // }
```
Comment (mongodben) [inline packages/mongodb-rag-ingest/src/embed/chunkCode.ts:58]:
```suggestion
```

File: packages/mongodb-rag-ingest/src/embed/chunkCode.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  scala: "scala",
+  swift: "swift",
+  latex: "latex",
+  html: "html",
+};
+
+export function isSupportedLanguage(
+  str: PageFormat
+): str is LangchainSupportedCodePageFormat {
+  return str in supportedLanguageMap;
```
Comment (mongodben) [inline packages/mongodb-rag-ingest/src/embed/chunkCode.ts:108]: what's the `in` operator here?

**Review by mongodben — APPROVED** [review]
LGTM.

next typescript level unlocked on this one

File: packages/mongodb-rag-ingest/src/embed/chunkCode.ts (hunk 4 above)
Comment (nlarew) [reply packages/mongodb-rag-ingest/src/embed/chunkCode.ts:108]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in

In this case it's equivalent to

```js
Object.keys(supportedLanguageMap).includes(str)
```

**Pull request merged**